yarn-project/p2p/src/service/reqresp/reqresp.ts

```diff
@@ -65,20 +74,33 @@ export class ReqResp {
    * @returns - The response from the peer, otherwise undefined
    */
   async sendRequest(subProtocol: ReqRespSubProtocol, payload: Buffer): Promise<Buffer | undefined> {
-    // Get active peers
-    const peers = this.libp2p.getPeers();
-
-    // Attempt to ask all of our peers
-    for (const peer of peers) {
-      const response = await this.sendRequestToPeer(peer, subProtocol, payload);
-
-      // If we get a response, return it, otherwise we iterate onto the next peer
-      // We do not consider it a success if we have an empty buffer
-      if (response && response.length > 0) {
-        return response;
+    const requestFunction = async () => {
+      // Get active peers
+      const peers = this.libp2p.getPeers();
+
+      // Attempt to ask all of our peers
+      for (const peer of peers) {
+        const response = await this.sendRequestToPeer(peer, subProtocol, payload);
```

Comment:
It took me a bit to appreciate that we ask the nodes serially, and so the collective timeout is somewhat dependent on the individual timeout.

Should the collective timeout be instead defined based on the number of peers we want to dial serially?
For example, the config in the current test implicitly caps us at 2 peers we can ask.

Also, while thinking of it, we'll want to track how often nodes are needing to ask for data, and how long it take them to get it. If the answer is "frequently" and "a long time" we may need to ask peers in parallel.

Reply:
Yes this is true, they are dependent on each other, further more these timeouts should probably be tuned to be per sub protocol, as the timings will be entirely different. 
Theres alot of complexity creeping up

Reply:
https://github.com/AztecProtocol/aztec-packages/issues/8458

Reply:
On the above, i began working on a rate limiting module today which will keep track of the who is asking for what. I'll have to integrate this data with the telemetry